processor/agentmetricsprocessor/utils_combine_process_metrics.go

```diff
@@ -0,0 +1,274 @@
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+)
+
+// The following code performs a translation from metrics that store process
+// information as resources to metrics that store process information as labels.
+//
+// Starting format:
+//
+//    ResourceMetrics         ResourceMetrics               ResourceMetrics
+// +-------------------+ +-----------------------+     +-----------------------+
+// |  Resource: Empty  | |  Resource: Process 1  |     |  Resource: Process X  |
+// +-------+---+-------+ +---------+---+---------+ ... +---------+---+---------+
+// |Metric1|...|MetricN| |MetricN+1|...|MetricN+M|     |MetricN+1|...|MetricN+M|
+// +-------+---+-------+ +---------+---+---------+     +---------+---+---------+
+//
+// Converted format:
+//
+//                             ResourceMetrics
+// +---------------------------------------------------------------------+
+// |                           Resource: Empty                           |
+// +-------+---+-------+----------------------+---+----------------------+
+// |Metric1|...|MetricN|MetricN+1{P1, ..., PX}|...|MetricN+M{P1, ..., PX}|
+// +-------+---+-------+----------------------+---+----------------------+
+//
+// Assumptions:
+// * There is at most one resource metrics slice without process resource info (will raise error if not)
+// * There is no other resource info supplied that needs to be retained (info may be silently lost if it exists)
+// * All metrics with the same name have the same descriptor (process metrics will be merged if the names match)
+
+func combineProcessMetrics(rms pdata.ResourceMetricsSlice) error {
+	// create collection of combined process metrics, disregarding any ResourceMetrics
+	// with no process resource attributes as "otherMetrics"
+	processMetrics, otherMetrics, err := createProcessMetrics(rms)
+	if err != nil {
+		return err
+	}
+
+	// if no non-process metrics were supplied, initialize an empty
+	// ResourceMetrics object
+	if otherMetrics.IsNil() {
+		otherMetrics.InitEmpty()
+		ilms := otherMetrics.InstrumentationLibraryMetrics()
+		ilms.Resize(1)
+		ilms.At(0).InitEmpty()
+	}
+
+	// append the process metrics to the other ResourceMetrics object
+	metrics := otherMetrics.InstrumentationLibraryMetrics().At(0).Metrics()
+	// ideally, we would Resize & Set, but function not available at this time
+	for _, metric := range processMetrics {
+		metrics.Append(&metric.Metric) // also, why does append take a pointer?
+	}
+
+	// TODO: This is super inefficient. Instead, we should just return a new
+	// data.MetricData struct, but currently blocked as it is internal
+	rms.Resize(1)
+	otherMetrics.CopyTo(rms.At(0))
+	return nil
+}
+
+func createProcessMetrics(rms pdata.ResourceMetricsSlice) (processMetrics convertedMetrics, otherMetrics pdata.ResourceMetrics, err error) {
+	processMetrics = convertedMetrics{}
+	otherMetrics = pdata.NewResourceMetrics()
```

Comment:
I think I would prefer this variable to be initialized as nil initially, since you never append data to it in this function

Reply:
I've generally avoided creating nil objects when dealing with `pdata` structs since internally these objects have double pointers to internal data.

An uninitialised pdata object (where `InitEmpty()` has not yet been called) is essentially equivalent to nil (as per the `IsNil()` function). This is not very idiomatic but seems to be the right way to deal with these objects. WDYT?